source/xlld/wrap.d

```diff
@@ -1266,28 +1294,139 @@ string wrapModuleFunctionStr(string moduleName, string funcName)(in string calli
     static assert(registerAttrs[0].argumentText.value == "Array to add");
 }
 
+void wrapAsync(alias F, A, T...)(ref A allocator, immutable XLOPER12 asyncHandle, T args) {
+    import xlld.framework: Excel12f;
+    import std.concurrency: spawn;
+    import std.format: format;
+
+    string toDArgsStr() {
+        import std.string: join;
+        import std.conv: text;
+
+        string[] pointers;
+        foreach(i; 0 .. T.length) {
+            pointers ~= text("&args[", i, "]");
+        }
+
+        return q{toDArgs!F(allocator, %s)}.format(pointers.join(", "));
+    }
+
+    mixin(q{alias DArgs = typeof(%s);}.format(toDArgsStr));
+    DArgs dArgs;
+
+    // Convert all Excel types to D types. This needs to be done here because the
+    // asynchronous part of the computation can't call back into Excel, and converting
+    // to D types requires calling coerce.
+    try {
+        mixin(q{dArgs = %s;}.format(toDArgsStr));
+    } catch(Throwable t) {
+        static if(isGC!F) {
+            import xlld.xll: log;
+            log("ERROR: Could not convert to D args for asynchronous function " ~
+                __traits(identifier, F));
+        }
+    }
+
+    try
+        spawn(&wrapAsyncImpl!(F, A, DArgs), allocator, asyncHandle, cast(immutable)dArgs);
+    catch(Exception ex) {
+        XLOPER12 functionRet, ret;
+        functionRet.xltype = XlType.xltypeErr;
+        Excel12f(xlAsyncReturn, &ret, cast(XLOPER12*)&asyncHandle, &functionRet);
+    }
+}
+
+void wrapAsyncImpl(alias F, A, T)(ref A allocator, XLOPER12 asyncHandle, T dArgs) {
+    import xlld.framework: Excel12f;
+    import xlld.xlcall: xlAsyncReturn;
+    import std.traits: Unqual;
+
+    // get rid of the temporary memory allocations for the conversions
+    scope(exit) freeDArgs(allocator, cast(Unqual!T)dArgs);
+
+    auto functionRet = callWrapped!F(dArgs);
+
+    XLOPER12 xl12ret;
+    const errorCode = () @trusted {
+        return Excel12f(xlAsyncReturn, &xl12ret, &asyncHandle, functionRet);
+    }();
+}
+
+// if a function is not @nogc, i.e. it uses the GC
+private bool isGC(alias F)() {
+    import std.traits: functionAttributes, FunctionAttribute;
+    enum nogc = functionAttributes!F & FunctionAttribute.nogc;
+    return !nogc;
+}
+
+// this has to be a top-level function and can't be declared in the unittest
+version(unittest) private double twice(double d) { return d * 2; }
+
+@("wrapAsync")
+@system unittest {
+    import xlld.test_util: lastAsyncReturn;
+    import std.datetime: Clock;
+    import core.thread;
+
+    const now = Clock.currTime;
```

Comment:
It would be better to use `core.time.MonoTime` for this. If the wall clock shifts at all while this is running, it might fail or run for a stupidly long time, depending on how the clock shifted, and the code would look pretty much the same except that it would have `MonoTime` instead of `Clock`.

Either way, I'd suggest renaming `now` to something like `start`, since you're basically subtracting `currTime` from `currTime` or `now` from `now`, which is a bit odd naming-wise, much as it isn't a big deal.